src/services/cohortCreation/index.ts

```diff
@@ -1,10 +1,120 @@
+import apiRequest from '../apiRequest'
 import apiBack from '../apiBackCohort'
 import { CONTEXT } from '../../constants'
 
-import { Cohort_Creation_API_Response } from 'types'
+import { Cohort_Creation_API_Response, FHIR_API_Response, Query, QueryGroup } from 'types'
+import { GroupTypeKind, IGroup, IGroup_Member, IPatient } from '@ahryman40k/ts-fhir-types/lib/R4'
+import { getApiResponseResources } from '../../utils/apiHelpers'
+
+const PATIENT_MAX_COUNT = 500
+
+const usePatientIdsByQueryMemo = () => {
+  const patientIdsByQuery: Record<string, (string | undefined)[] | undefined> = {}
+  return async (query: string): Promise<(string | undefined)[]> => {
+    if (!query) return []
+    if (!patientIdsByQuery[query]) {
+      const response = await apiFhir.get<FHIR_API_Response<IPatient>>(query)
+      const patients = getApiResponseResources(response)
+      patientIdsByQuery[query] = patients ? patients.map((patient) => patient.id) : []
+    }
+    return patientIdsByQuery[query] ?? []
+  }
+}
+
+const getPatients = usePatientIdsByQueryMemo()
+
+const createCohortGroup = async (json_query: string): Promise<IGroup> => {
+  const query: Query = JSON.parse(json_query)
+  const perimeters = query.sourcePopulation.caresiteCohortList
+  const patientFilter = `_count=${PATIENT_MAX_COUNT}&_has:Encounter:subject:service-provider=${perimeters.join(',')}`
+  const criteriaGroup = query.request[0].criteria
+
+  const aggregatePatients = async (
+    group: QueryGroup,
+    patientIds: (string | undefined)[],
+    currentCohort: (string | undefined)[]
+  ): Promise<(string | undefined)[]> => {
+    switch (group._type) {
+      case 'andGroup':
+        if (group.isInclusive)
+          return currentCohort.length === 0 ? [...patientIds] : intersection(currentCohort, patientIds)
+        else
+          return currentCohort.length === 0
+            ? [...pullAll(await getPatients(`/Patient?${patientFilter}`), patientIds)]
+            : [...pullAll(currentCohort, patientIds)]
+      case 'orGroup':
+        return group.isInclusive
+          ? union(currentCohort, patientIds)
+          : union(currentCohort, pullAll(await getPatients(`/Patient?${patientFilter}`), patientIds))
+      default:
+        return [...currentCohort]
+    }
+  }
+
+  const queryByResourceType = (criteria: QueryGroup): string => {
+    switch (criteria.resourceType) {
+      case 'Patient':
+        return `/Patient?${patientFilter}&${criteria.filterSolr}`
+      case 'Condition':
+        return `/Patient?${patientFilter}&_has:Condition:patient:${criteria.filterSolr}`
+      default:
+        return ''
+    }
+  }
+
+  const queries = criteriaGroup.reduce<string[]>((acc, group) => {
+    return [...acc, ...group.criteria.reduce<string[]>((acc, criteria) => [...acc, queryByResourceType(criteria)], [])]
+  }, [])
+  const fetchPatients = async (queries: string[]) => await Promise.all(queries.map((query) => getPatients(query)))
+  await fetchPatients(queries)
```

Comment:
I don't get what we really do here.

Why instantiating `fetchPatients` instead of doing
```suggestion
  await Promise.all(queries.map((query) => getPatients(query)))
```

And why the promises values are not kept..? Is it just to refresh the `usePatientIdsByQueryMemo` cache ?